internal/driver/incominglistener.go

```diff
@@ -7,55 +7,92 @@
+		driver.Logger.Errorf("[Incoming listener] Incoming data ignored, incoming topic data should have format .../<device_name>/<source_name>: `%s`", incomingTopic)
 		return
 	}
 	deviceName = metaData[0]
-	resourceName = metaData[1]
-	reading = string(message.Payload())
+	sourceName = metaData[1]
 
-	deviceObject, ok := d.sdk.DeviceResource(deviceName, resourceName)
+	deviceCommand, ok := d.sdk.DeviceCommand(deviceName, sourceName)
 	if !ok {
-		driver.Logger.Errorf("[Incoming listener] Incoming reading ignored, device resource `%s` not found from the device `%s`", resourceName, deviceName)
-		return
-	}
+		deviceResource, ok := d.sdk.DeviceResource(deviceName, sourceName)
+		if !ok {
+			driver.Logger.Errorf("[Incoming listener] Incoming data ignored, source name `%s` not found as Device Command or Device Resource on the device `%s`", sourceName, deviceName)
+			return
+		}
+
+		if !strings.Contains(strings.ToUpper(deviceResource.Properties.ReadWrite), "R") {
+			driver.Logger.Errorf("[Incoming listener] Incoming data ignored, Device Resource `%s` not Readable", sourceName)
+			return
+		}
+
+		deviceResources = append(deviceResources, deviceResource)
+	} else {
+		if !strings.Contains(strings.ToUpper(deviceCommand.ReadWrite), "R") {
+			driver.Logger.Errorf("[Incoming listener] Incoming data ignored, Device Command `%s` not Readable", sourceName)
+			return
+		}
+
+		for _, resourceOperation := range deviceCommand.ResourceOperations {
+			deviceResource, ok := d.sdk.DeviceResource(deviceName, resourceOperation.DeviceResource)
+			if !ok {
+				driver.Logger.Errorf("[Incoming listener] Incoming data ignored, resource name `%s` from Device Command %s not found on the device `%s`", resourceOperation.DeviceResource, sourceName, deviceName)
+				return
+			}
 
-	req := models.CommandRequest{
-		DeviceResourceName: resourceName,
-		Type:               deviceObject.Properties.ValueType,
+			deviceResources = append(deviceResources, deviceResource)
+		}
 	}
 
-	result, err := newResult(req, reading)
+	var asyncData map[string]interface{}
+	err := json.Unmarshal(message.Payload(), &asyncData)
 	if err != nil {
-		driver.Logger.Errorf("[Incoming listener] Incoming reading ignored, %v", err)
+		driver.Logger.Errorf("[Incoming listener] Error un-marshaling incoming data : %v", err)
 		return
 	}
 
-	asyncValues := &models.AsyncValues{
+	var commandValues []*sdkModels.CommandValue
+
+	for _, resource := range deviceResources {
+		asyncValue, ok := asyncData[resource.Name]
+		if !ok {
+			driver.Logger.Errorf("[Incoming listener] Incoming data ignored: Resource Name %s not found in payload (%s)", resource.Name, string(message.Payload()))
+			return
+		}
+
+		commandValue, err := newResult(resource, asyncValue)
+		if err != nil {
+			driver.Logger.Errorf("[Incoming listener] Incoming data ignored: %v", err)
```

Comment:
```suggestion
			driver.Logger.Errorf("[Incoming listener] Incoming data ignored: got error %v", err)
```

Reply:
`newResult` adds to context to the error, so no additional context needed here.